File: zig/isbn-verifier/isbn_verifier.zig
pub fn isValidIsbn10(s: []const u8) bool {
    var sum: u16 = 0;

    var i: u8 = 0;
    for (s) |c| {
        var digit: u8 = switch (c) {
            '0'...'9' => c - '0',
            'X' => if (i == 9) 10 else return false,
            '-' => continue,
            else => return false,
        };

        sum += (10 - i) * digit;
        i += 1;
    }

    return i == 10 and sum % 11 == 0;
}

File: python/isbn-verifier/isbn_verifier.py
from string import digits

valid_characters = set(digits).union({"-", "X"})


def is_valid(isbn: str):
    if any(char not in valid_characters for char in isbn):
        return False

    digits = [10 if digit == "X" else int(digit) for digit in isbn if digit != "-"]

    return (
        len(digits) == 10
        and all((digit != 10 or index == 9) for index, digit in enumerate(digits))
        and sum((digit * (10 - index)) for index, digit in enumerate(digits)) % 11 == 0
    )
